esp-hal/src/rmt.rs

```diff
@@ -250,59 +250,235 @@ pub enum Error {
+    pub const fn try_new(level1: Level, length1: u16, level2: Level, length2: u16) -> Option<Self> {
+        if length1 >= 0x8000 || length2 >= 0x8000 {
+            return None;
+        }
+
+        // SAFETY:
+        // - We just checked that length1 and length2 have their MSB cleared.
+        Some(unsafe { Self::new_unchecked(level1, length1, level2, length2) })
+    }
+
+    /// Create a new empty instance.
+    ///
+    /// This corresponds to the all-zero `PulseCode`, i.e. an end marker with
+    /// level `Level::Low`.
+    #[inline]
+    pub const fn empty() -> Self {
+        Self(0)
+    }
+
+    /// Create a new instance that is and end marker with `Level::Low`.
+    ///
+    /// This is equivalent to `PulseCode::empty()` and provided as a more
+    /// semantic alias.
+    // FIXME: Consider adding a variant with `level1`, `length1` and `level2` arguments
+    // which sets `length2 = 0` so that it is still guaranteed to return an end
+    // marker.
+    #[inline]
+    pub const fn end_marker() -> Self {
+        Self(0)
+    }
+
+    /// Set all levels and lengths to 0.
+    ///
+    /// In other words, assigns the value of `PulseCode::empty()` to `self`.
+    #[inline]
+    pub fn reset(&mut self) {
+        self.0 = 0
+    }
+
+    /// Logical output level in the first pulse code interval
+    #[inline]
+    pub const fn level1(&self) -> Level {
+        // Can't use Level::from(bool) since it is non-const
+        if 0 != (self.0 & (1 << (LEVEL1_SHIFT + 15))) {
+            Level::High
+        } else {
+            Level::Low
+        }
+    }
 
     /// Logical output level in the second pulse code interval
-    fn level2(&self) -> Level;
+    #[inline]
+    pub const fn level2(&self) -> Level {
+        // Can't use Level::from(bool) since it is non-const
+        if 0 != (self.0 & (1 << (LEVEL2_SHIFT + 15))) {
+            Level::High
+        } else {
+            Level::Low
+        }
+    }
+
+    /// Length of the first pulse code interval (in clock cycles)
+    #[inline]
+    pub const fn length1(&self) -> u16 {
+        ((self.0 >> LEVEL1_SHIFT) & 0x7FFF) as u16
+    }
 
     /// Length of the second pulse code interval (in clock cycles)
-    fn length2(&self) -> u16;
-}
+    #[inline]
+    pub const fn length2(&self) -> u16 {
+        ((self.0 >> LEVEL2_SHIFT) & 0x7FFF) as u16
+    }
 
-impl PulseCode for u32 {
-    fn new(level1: Level, length1: u16, level2: Level, length2: u16) -> Self {
-        let level1 = ((bool::from(level1) as u32) << 15) | (length1 as u32 & 0b111_1111_1111_1111);
-        let level2 = ((bool::from(level2) as u32) << 15) | (length2 as u32 & 0b111_1111_1111_1111);
-        level1 | (level2 << 16)
+    /// Set level1
+    #[inline]
+    pub const fn with_level1(mut self, level: Level) -> Self {
+        if matches!(level, Level::High) {
+            self.0 |= 1 << (LEVEL1_SHIFT + 15);
+        } else {
+            self.0 &= !(1 << (LEVEL1_SHIFT + 15));
+        }
+        self
```

Comment:
Could you deduplicate this a bit? We don't need 4 separate implementations of the same thing (with a different shifting factor).

You could also precompute 1<<(SHIFT+15) into its own constant, I think.

Reply:
I'm not sure I understand what you suggest to deduplicate here. That said, I've simplified this by precomputing a few more `*_SHIFT` and `*_MASK` constants and by making better use of converting `Level` to `bool`.

I'll mark this as resolved for now, let me know if you see further potential for deduplication once I push an updated branch.